tests/integration-tests/src/test/java/org/apache/activemq/artemis/tests/integration/amqp/connect/AMQPConnectSaslTest.java

```diff
@@ -0,0 +1,299 @@
+      assertTrue("Broker did not open connection in alotted time", awaitConnectionOpen);
+
+      assertEquals(PLAIN, authenticator.getChosenMech());
+      assertArrayEquals(expectedPlainInitialResponse(USER, PASSWD), authenticator.getInitialResponse());
+   }
+
+   @Test(timeout = 20000)
+   public void testConnectsWithExternal() throws Exception {
+      doConnectWithExternalTestImpl(true);
+   }
+
+   @Test(timeout = 20000)
+   public void testExternalIgnoredWhenNoClientCertSupplied() throws Exception {
+      doConnectWithExternalTestImpl(false);
+   }
+
+   private void doConnectWithExternalTestImpl(boolean requireClientCert) throws ExecutionException, InterruptedException, Exception {
+      CountDownLatch serverConnectionOpen = new CountDownLatch(1);
+      // The test server always offers EXTERNAL, i.e sometimes mistakenly, to verify that the broker only selects it when it actually
+      // has a client-cert. Real servers shouldnt actually offer the mechanism to a client that didnt have to provide a cert.
+      TestAuthenticator authenticator = new TestAuthenticator(true, EXTERNAL, PLAIN);
+
+      final String keyStorePath = this.getClass().getClassLoader().getResource(SERVER_KEYSTORE_NAME).getFile();
+      JksOptions jksKeyStoreOptions = new JksOptions().setPath(keyStorePath).setPassword(SERVER_KEYSTORE_PASSWORD);
+
+      ProtonServerOptions serverOptions = new ProtonServerOptions();
+      serverOptions.setSsl(true);
+      serverOptions.setKeyStoreOptions(jksKeyStoreOptions);
+
+      if (requireClientCert) {
+         final String trustStorePath = this.getClass().getClassLoader().getResource(TRUSTSTORE_NAME).getFile();
+         JksOptions jksTrustStoreOptions = new JksOptions().setPath(trustStorePath).setPassword(TRUSTSTORE_PASSWORD);
+
+         serverOptions.setTrustStoreOptions(jksTrustStoreOptions);
+         serverOptions.setClientAuth(ClientAuth.REQUIRED);
+      }
+
+      mockServer = new MockServer(vertx, serverOptions, () -> authenticator, serverConnection -> {
+         serverConnection.openHandler(serverSender -> {
+            serverConnectionOpen.countDown();
+            serverConnection.closeHandler(x -> serverConnection.close());
+            serverConnection.open();
+         });
+      });
+
+      String amqpServerConnectionURI = "tcp://localhost:" + mockServer.actualPort() +
+            "?sslEnabled=true;trustStorePath=" + TRUSTSTORE_NAME + ";trustStorePassword=" + TRUSTSTORE_PASSWORD;
+      if (requireClientCert) {
+         amqpServerConnectionURI += ";keyStorePath=" + CLIENT_KEYSTORE_NAME + ";keyStorePassword=" + CLIENT_KEYSTORE_PASSWORD;
+      }
+
+      AMQPBrokerConnectConfiguration amqpConnection = new AMQPBrokerConnectConfiguration("testSimpleConnect", amqpServerConnectionURI);
+      amqpConnection.setReconnectAttempts(0);// No reconnects
+      amqpConnection.setUser(USER); // Wont matter if EXTERNAL is offered and a client-certificate is provided, but will otherwise.
```

Comment:
shouldn't we add actual java properties supporting this?

I mean, adding properties on the .XSD? 

I know you may say you could go all the way to a lot of stuff that's not he URI, but this seems a nice use case to support through distinct properties on the AMQPBrokerConnection configuration.

Reply:
The test was just checking it uses external if offered, even in the face of other options, rather than because its the only option. There isnt any config needed for it, just basically 'it is offered, I have a cert'.

I did think about adding the ability to restrict which mechanisms it can pick, which would make sense, but I decided its a seperate change. Also, its more likely to make the StackOverflowError occur so I think fixing that would need to occur concurrently or before.